adapters/repos/db/vector/hnsw/sharded_node_locks.go

```diff
@@ -0,0 +1,66 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package hnsw
+
+import "sync"
+
+const nodeLockStripe = uint64(512)
+
+type shardedNodeLocks []*sync.RWMutex
+
+func newShardedNodeLocks() shardedNodeLocks {
+	snl := make([]*sync.RWMutex, nodeLockStripe)
+	for i := range snl {
+		snl[i] = new(sync.RWMutex)
+	}
+	return snl
+}
+
+func (snl shardedNodeLocks) LockAll() {
```

Comment:
nitpicking (feel free to ignore)
Since we now have a nice abstraction, should we instead have a master lock instead of locking all the locks?
Locking / unlocking 512 locks takes about 3 microseconds on my machine, so I'm not sure if my suggestion is useful as it only matters if this function is called on a hot path.

Reply:
I like the concept of master lock, though I don't have idea how to implement it to be dependent (bidirectionally) on individual locks.

Master lock should wait until all individual one are unlocked, individual lock should wait for master to be unlocked.
It gets even more complicated with rlocking.
Happy to change that, but need some conceptual help.

Reply:
I had an idea with a master RWMutex, where `RLock` would be used for individual locks and `Lock` would be used for `LockAll`, but it doesn't work with rlocking.
I was looking for a quick change, but it's getting too complex and I'm not sure this is worth it or even worth slowing down this PR.